internal/controller/stream_controller.go

```diff
@@ -18,39 +18,371 @@ package controller
+	}
+
+	if !stream.Spec.PreventDelete && !r.ReadOnly() {
+		log.Info("Deleting stream.")
+		err := r.WithJetStreamClient(streamConnOpts(stream.Spec), func(js jetstream.JetStream) error {
+			return js.DeleteStream(ctx, stream.Spec.Name)
+		})
+		if errors.Is(err, jetstream.ErrStreamNotFound) {
+			log.Info("Managed stream was already deleted.")
+		} else if err != nil {
+			return fmt.Errorf("delete stream during finalization: %w", err)
+		}
+	} else {
+		log.Info("Skipping stream deletion.",
+			"streamName", stream.Spec.Name,
+			"preventDelete", stream.Spec.PreventDelete,
+			"read-only", r.ReadOnly(),
+		)
+	}
+
+	log.Info("Removing stream finalizer.")
+	if ok := controllerutil.RemoveFinalizer(stream, streamFinalizer); !ok {
+		return errors.New("failed to remove stream finalizer")
+	}
+	if err := r.Update(ctx, stream); err != nil {
+		return fmt.Errorf("remove finalizer: %w", err)
+	}
+
+	return nil
+}
+
+func (r *StreamReconciler) createOrUpdate(ctx context.Context, log logr.Logger, stream *api.Stream) error {
+
+	// Create or Update the stream based on the spec
+	if stream.Spec.PreventUpdate || r.ReadOnly() {
+		log.Info("Skipping stream creation or update.",
+			"streamName", stream.Spec.Name,
+			"preventDelete", stream.Spec.PreventDelete,
+			"read-only", r.ReadOnly(),
+		)
+		return nil
+	}
+
+	// Map spec to stream targetConfig
+	targetConfig, err := streamSpecToConfig(&stream.Spec)
+	if err != nil {
+		return fmt.Errorf("map spec to stream targetConfig: %w", err)
+	}
+
+	// CreateOrUpdateStream is called on every reconciliation when the stream is not to be deleted.
+	// TODO(future-feature): Do we need to check if config differs?
+	err = r.WithJetStreamClient(streamConnOpts(stream.Spec), func(js jetstream.JetStream) error {
+		log.Info("create or update stream", "streamName", targetConfig.Name)
+		_, err = js.CreateOrUpdateStream(ctx, targetConfig)
+		return err
+	})
+	if err != nil {
+		err = fmt.Errorf("create or update stream: %w", err)
+		stream.Status.Conditions = updateReadyCondition(stream.Status.Conditions, v1.ConditionFalse, "Errored", err.Error())
+		if err := r.Status().Update(ctx, stream); err != nil {
+			log.Error(err, "Failed to update ready condition to Errored.")
+		}
+		return err
+	}
+
+	// update the observed generation and ready status
+	stream.Status.ObservedGeneration = stream.Generation
+	stream.Status.Conditions = updateReadyCondition(
+		stream.Status.Conditions,
+		v1.ConditionTrue,
+		"Reconciling",
+		"Stream successfully created or updated.",
+	)
+	err = r.Status().Update(ctx, stream)
+	if err != nil {
+		return fmt.Errorf("update ready condition: %w", err)
+	}
+
+	return nil
+}
+
+// streamConnOpts extracts nats connection relevant fields from the given stream spec as connectionOptions.
+func streamConnOpts(spec api.StreamSpec) *connectionOptions {
+	return &connectionOptions{
+		Account: spec.Account, // TODO(review): Where does Spec.Account have to be considered?
```

Comment:
The account here is relevant to what account the Stream should be created in. creds/nkey will be inherently associated with an account, so if we are using `CRDConnect` then the auth provided in the spec should match that account.

It seems potentially redundant to have the `Account` option since you would need to provide different auth options if you are changing from the base config, and the account will be inferred. I'll double check on this.

Reply:
Added a note in the base jetstream controller, where creds and nkey are handled that the account field exists but is not used.